Nestedness

slevin75 opened the pull request:
This function will find all the upstream gages of each gage in gagesii.  The output is a data frame where the column name is the downstream gage and any row that has a 1 is a nested upstream gage.  The gage itself will be listed as a nested gage, so every column should have at least one 1.  It is set up to run for all of gagesii but if you want to be able to subset it by region, I can add an argument to subset it by the screened site list first.  

I hand checked about 35 sites to make sure it was giving accurate results and they were all fine.  The only problem I found was that there are 5 sites for which the navigate_nldi function in nhdplusTools throws an error where no data is returned (gages are:  02084557, 09406300,  09512200, 10143500, 10172200).  I am wondering if those gages have an incorrect comid but I don't know how to check this.  I still can't figure out how to get the comid from the nhdplus.  I don't see any attributes that match it.  

I do have one small concern because this is currently set up to run the navigate_nldi function in a loop so it repeatedly calls on that web service and I'm afraid it might time out sometimes. It did it to me one time only, so if it seems like a problem, I guess we can talk about that.

jds485 commented:
> I am wondering if those gages have an incorrect comid but I don't know how to check this. I still can't figure out how to get the comid from the nhdplus. I don't see any attributes that match it.

@ajsekell do you have suggestions for checking the COMID of these gages? 
<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/jsmith/AppData/Local/Temp/5/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/jsmith/AppData/Local/Temp/5/msohtmlclip1/01/clip_filelist.xml">

<!--table
	{mso-displayed-decimal-separator:"\.";
	mso-displayed-thousand-separator:"\,";}
@page
	{margin:1.0in .75in 1.0in .75in;
	mso-header-margin:.5in;
	mso-footer-margin:.5in;}
tr
	{mso-height-source:auto;}
col
	{mso-width-source:auto;}
br
	{mso-data-placement:same-cell;}
td
	{padding-top:1px;
	padding-right:1px;
	padding-left:1px;
	mso-ignore:padding;
	color:black;
	font-size:12.0pt;
	font-weight:400;
	font-style:normal;
	text-decoration:none;
	font-family:Calibri, sans-serif;
	mso-font-charset:0;
	mso-number-format:General;
	text-align:general;
	vertical-align:bottom;
	border:none;
	mso-background-source:auto;
	mso-pattern:auto;
	mso-protection:locked visible;
	white-space:nowrap;
	mso-rotate:0;}
.xl65
	{background:yellow;
	mso-pattern:black none;}
-->

</head>

<body link="#0563C1" vlink="#954F72">



ID | STANAME | LAT | LON | COMID
-- | -- | -- | -- | --
T02084557 | VAN SWAMP NEAR HOKE, NC | 35.730833 | -76.746111 | 933020168
T09406300 | KANARRA CREEK AT KANARRAVILLE, UTAH | 37.538034 | -113.168561 | 10020312
T09512200 | SALT RIVER TRIB. IN S. MNT. PARK, PHOENIX, ARIZ. | 33.34699 | -112.08487 | 20478504
T10143500 | CENTERVILLE CREEK ABV. DIV NEAR CENTERVILLE, UT | 40.916334 | -111.862993 | 10277294
T10172200 | RED BUTTE CREEK AT FORT DOUGLAS, NEAR SLC, UT | 40.779946 | -111.806045 | 10389356



</body>

</html>

ajsekell commented:
Since it's only 5 sites, I'll do it the old fashioned way in ArcMap.

Andrew Sekellick
Physical Scientist
USGS MD-DE-DC Water Science Center
443-498-5580
***@***.******@***.***>

________________________________
From: Jared D Smith ***@***.***>
Sent: Friday, February 4, 2022 10:21 AM
To: USGS-R/regional-hydrologic-forcings-ml ***@***.***>
Cc: Sekellick, Andrew J ***@***.***>; Mention ***@***.***>
Subject: [EXTERNAL] Re: [USGS-R/regional-hydrologic-forcings-ml] Nestedness (PR #49)




 This email has been received from outside of DOI - Use caution before clicking on links, opening attachments, or responding.



I am wondering if those gages have an incorrect comid but I don't know how to check this. I still can't figure out how to get the comid from the nhdplus. I don't see any attributes that match it.

@ajsekell<https://gcc02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fajsekell&data=04%7C01%7Cajsekell%40usgs.gov%7C8befb72c9b5141b8154c08d9e7f2141b%7C0693b5ba4b184d7b9341f32f400a5494%7C0%7C0%7C637795849183186366%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000&sdata=OLosdgt0akG0e7VBFsCiOsBg4eC9ryoWLN9F6Cpjjgg%3D&reserved=0> do you have suggestions for checking the COMID of these gages?

ID      STANAME LAT     LON     COMID
T02084557       VAN SWAMP NEAR HOKE, NC 35.730833       -76.746111      933020168
T09406300       KANARRA CREEK AT KANARRAVILLE, UTAH     37.538034       -113.168561     10020312
T09512200       SALT RIVER TRIB. IN S. MNT. PARK, PHOENIX, ARIZ.        33.34699        -112.08487      20478504
T10143500       CENTERVILLE CREEK ABV. DIV NEAR CENTERVILLE, UT 40.916334       -111.862993     10277294
T10172200       RED BUTTE CREEK AT FORT DOUGLAS, NEAR SLC, UT   40.779946       -111.806045     10389356

—
Reply to this email directly, view it on GitHub<https://gcc02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FUSGS-R%2Fregional-hydrologic-forcings-ml%2Fpull%2F49%23issuecomment-1030085191&data=04%7C01%7Cajsekell%40usgs.gov%7C8befb72c9b5141b8154c08d9e7f2141b%7C0693b5ba4b184d7b9341f32f400a5494%7C0%7C0%7C637795849183186366%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000&sdata=Dd9spAKUN6YWAeGF%2FpN6hYsVTRRVL53cTMLuqGLtzhE%3D&reserved=0>, or unsubscribe<https://gcc02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAV5VDZO44QNNKMFZJEUHMODUZPVIJANCNFSM5NSAUZ2A&data=04%7C01%7Cajsekell%40usgs.gov%7C8befb72c9b5141b8154c08d9e7f2141b%7C0693b5ba4b184d7b9341f32f400a5494%7C0%7C0%7C637795849183186366%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000&sdata=jQhpRbnJ1i0nhxJPeNcO9G43j%2FeJtUE3h%2B6qTs7p%2B9s%3D&reserved=0>.
Triage notifications on the go with GitHub Mobile for iOS<https://gcc02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fapps.apple.com%2Fapp%2Fapple-store%2Fid1477376905%3Fct%3Dnotification-email%26mt%3D8%26pt%3D524675&data=04%7C01%7Cajsekell%40usgs.gov%7C8befb72c9b5141b8154c08d9e7f2141b%7C0693b5ba4b184d7b9341f32f400a5494%7C0%7C0%7C637795849183186366%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000&sdata=CEJzGFmqpY8mDm0P7mN%2BOJ5JnX6fwGmviP8ATnqwiks%3D&reserved=0> or Android<https://gcc02.safelinks.protection.outlook.com/?url=https%3A%2F%2Fplay.google.com%2Fstore%2Fapps%2Fdetails%3Fid%3Dcom.github.android%26referrer%3Dutm_campaign%253Dnotification-email%2526utm_medium%253Demail%2526utm_source%253Dgithub&data=04%7C01%7Cajsekell%40usgs.gov%7C8befb72c9b5141b8154c08d9e7f2141b%7C0693b5ba4b184d7b9341f32f400a5494%7C0%7C0%7C637795849183186366%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000&sdata=Cww4qXBqQ69LjS1GIpPmC7%2BAYW4rMoVS%2FjlrScTP3Mc%3D&reserved=0>.
You are receiving this because you were mentioned.Message ID: ***@***.***>

jds485 commented:
Great, thanks! You can make edits to the COMID for these sites in the SharePoint General/Data/Gages2.1_RefSiteList.xlsx spreadsheet

jds485 commented:
> I'm afraid it might time out sometimes

We can implement the retry function for this case. Example: 
```
retry(expr = <navigate_nldi function>,
                   when = <some error message>,
                   max_tries = 10)
```

I'll look at the function code to see if there are alternatives

jds485 commented on _targets.R:
```diff
@@ -17,6 +17,7 @@ source("1_fetch/src/get_nwis_data.R")
 source("1_fetch/src/calc_HIT.R")
 source("1_fetch/src/calc_FDC.R")
 source("1_fetch/src/moving_window_functions.R")
+source("1_fetch/src/cross_validation_functions.R")
```
Let's number this as `4_setup_crossval` and use p4 for the targets names. I think we'll add cross validation functions to this later to select CV groups

slevin75 replied:
Want this in a new 4_crossval folder, correct? Not in the fetch directory?

jds485 replied:
yes, thanks!

jds485 commented on 1_fetch/src/cross_validation_functions.R:
```diff
@@ -0,0 +1,21 @@
+  
+   #assigns a 1 if the row name gage is upstream of the column name gage
+  upstream_df <- data.frame(matrix(ncol = length(gage_IDs), nrow = length(gage_IDs)))
+  colnames(upstream_df) <- gage_IDs; rownames(upstream_df) <- gage_IDs
```
`colnames(upstream_df) <- rownames(upstream_df) <- gage_IDs`

jds485 commented on 1_fetch/src/cross_validation_functions.R:
```diff
@@ -0,0 +1,21 @@
+   #assigns a 1 if the row name gage is upstream of the column name gage
+  upstream_df <- data.frame(matrix(ncol = length(gage_IDs), nrow = length(gage_IDs)))
+  colnames(upstream_df) <- gage_IDs; rownames(upstream_df) <- gage_IDs
+  upstream_df[,]<-0
```
instead, in the `matrix` function above you can specify `data = 0`

jds485 commented on 1_fetch/src/cross_validation_functions.R:
```diff
@@ -0,0 +1,21 @@
+  upstream_df <- data.frame(matrix(ncol = length(gage_IDs), nrow = length(gage_IDs)))
+  colnames(upstream_df) <- gage_IDs; rownames(upstream_df) <- gage_IDs
+  upstream_df[,]<-0
+  for (i in 1:length(gage_IDs)){
+    flowline <- navigate_nldi(list(featureSource = "comid", 
+                                       featureID = COMIDs[i]), 
+                                  mode = "upstreamTributaries",
+                                  distance_km=nav_distance_km)
+    upstream_list <- flowline$UT_flowlines$nhdplus_comid
+    upstream_df[which(COMIDs %in% upstream_list), i] <- 1
+    message(gage_IDs[i])
+  }
```
It would be helpful to report other information about nestedness to help us choose a threshold for how nested is too much for gages to be in a separate group. Does nhdplusTools have a function for pulling the upstream catchment area? Reporting the proportion overlap would be helpful. The scripts I sent you from the Drought project might do some of this because theirs is based on area

slevin75 replied:
I can take a look at that script and see what nhdplusTools has but I wonder if it might be easier to just use dataRetreival to get the drainage areas. Then, instead of just a 1/0 in the matrix, I could put the drainage area ratio upstream/downstream in the matrix instead. Let me know what you think of that idea.

jds485 replied:
oh that's definitely going to be faster! Good idea

jds485 reviewed (changes requested):
Looks good! I haven't run yet (I'll wait for Andrew to update the COMIDs). I have some change requests in the comments. We can save the comment about adding other nestedness info for Monday's meeting. If you have time, could you check if the Drought project code can be used for the catchment area calculation?

jds485 commented:
@slevin75 did you receive messages like this one when you ran the function? Several sites displayed this message for me.

`No data returned for: https://labs.waterdata.usgs.gov/api/nldi/linked-data/comid/10020312/navigation/UT/flowlines?distance=4500FALSE`

slevin75 commented:
> 

That was the error I got for those 5 sites where the navigate_nldi didn't work.  Looks like that one is for gage 09406300 which has com id 10020312

slevin75 commented:
no, I didn't check for that.  Although in other cases, it also returns the actual reach that the gage is on, so I feel like it should at least return one reach, being the one that the gage is on.

jds485 commented:
Okay, I see why you thought the COMIDs could be incorrect

slevin75 commented:
> There are sites for which the ratio of drainage areas is greater than 1

Yeah - I added the drainage area stuff at the end of the day on Friday and haven't double checked it all, so there could be some new problems I added in when I did that.

ajsekell commented:
02084557	VAN SWAMP NEAR HOKE, NC
^ this one is a canal/ditch (also a swamp). there is a flowline with that COMID but no corresponding catchment.

10143500 CENTERVILLE CREEK ABV. DIV NEAR CENTERVILLE, UT
^ this one is listed as a pipeline. no corresponding catchment.

10172200 RED BUTTE CREEK AT FORT DOUGLAS, NEAR SLC, UT
^this is a stream but it is disconnected from main network (catchment is a sink?).

09406300 KANARRA CREEK AT KANARRAVILLE, UTAH
^same as above...headwaters.

09512200 SALT RIVER TRIB. IN S. MNT. PARK, PHOENIX, ARIZ.
^same as above


Some of these were tricky. My advice would be to probably not use these these gages.

jds485 commented on 4_setup_crossval/src/cross_validation_functions.R:
```diff
@@ -0,0 +1,41 @@
+    ##gages located on any upstream COMID (includes the COMID of the downstream gage)
+    upstream_gages<-gage_IDs[which(COMIDs %in% upstream_list)]
+    
+    ##if there are more upstream gages than there are unique COMIDs, then 
+    ##there are more than one gage on a COMID.  If this happens, remove any
+    ##gage from the upstream gages list that has a larger drainage area than the
+    ##downstream_da.
```
Just to check my understanding, is this happening only when the "downstream" gage is on the same COMID as an "upstream" gage so that "upstream" gage is actually downstream?

slevin75 replied:
Yes.  I'm pretty sure this was the cause of the NAs as well. I'm running it again now to check this. Basically, I had been linking everything by COMID, thinking that was unique to each gage but in cases where two gages on one comid occured, it would include that larger gage as an upstream gage of the smaller gage.

jds485 replied:
haha, no worries, it was easy to go through. I'm hoping this will resolve the ratios that were a little greater than 1, but it may not have been the cause for the ratio that was ~5

slevin75 replied:
yes it was - that was the first one I looked at.  Basically there were two gages on that reach - one was ~4.5 sq miles and the other was 27 sq miles.

slevin75 commented:
OK, so the NAs in the nested gages matrix are because there are 13 gages for which there is no drainage area returned from NWIS.   Should I just exclude those or look for another way to get drainage area?

      gage_IDs    drainArea
336   02047783        NA
423  021973515        NA
473   02298554        NA
482   02312675        NA
501   02326526        NA
1108  06190540        NA
1248  06781600        NA
1692  09383409        NA
1711  09418700        NA
1821  10317480        NA
1857  11073001        NA
2156  13310800        NA
2157  13310850        NA

jds485 commented:
Based on the [description](https://waterdata.usgs.gov/wa/nwis/current?type=basinda) in NWIS for why sites would be missing DA data, it sounds like we can drop them. "In some cases the drainage area will be reported as either "unknown" or "undetermined". This means either the drainage hasn't been determined, common with new stations where we will determine the area and report it future publications, or it means the drainage area has no relative meaning, where the station is a diversion, return, irrigation canal, lake, or similar hydrologic features where the streamflow or lake contents can not be related to a river basin and its drainage area."

jds485 commented on 1_fetch/src/cross_validation_functions.R:
```diff
@@ -0,0 +1,21 @@
+
+get_nested_gages<-function(gagesii,nav_distance_km){
```
Can you move .R file with this function to the `3_` directory? `git mv` can be used to move while git tracking the move

jds485 replied:
Oh wait, it might already be there? I see the same .R file name in 4_

jds485 replied:
Did you `git rm` the 1_fetch version of this file? I still see it in my files after pulling the latest commits.

slevin75 commented:
@jds485 I think I have addressed everything now, so it should be ready to re-review.

jds485 commented on 1_fetch/src/get_nwis_data.R:
```diff
@@ -1,11 +1,15 @@
   pk_screen <- whatNWISdata(siteNumber=site_nums, service="pk", convertType = FALSE)
   sites_with_data <- intersect(dv_screen$site_no, pk_screen$site_no)
+  #remove sites that do not have a drainage area in NWIS
+  rm.sites<-which(is.na(readNWISsite(siteNumbers= sites_with_data)$drain_area_va))
```
change to `rm_sites` to keep with naming convention

jds485 commented:
Did you drop the 5 sites Andrew cross-checked and suggested to drop? I still see "no data returned" messages
